### ISSUE TITLE
(To be Verified) Calico selecting wrong device for deletion during cleanup

### ISSUE THREAD

**Issue by andrewhowdencom** [projectcalico/calico#3960]
Hola crew, 

While debugging a pod that apparently stopped receiving traffic for an unknown reason today and had to be flushed, I discovered that the pod was using a network device that apparently had been incorrectly cleaned up by calico as part of another pod deletion.

Note: My networking knowledge is ... okay. Much of the specifics are over my head.

## Expected Behavior

1. A new pod is scheduled
2. That new pod starts the pause container and establishes the network namespace (or so)
3. Calico is called via cni to create the necessary adapters and so
4. Those adapters are created successful, and kubernetes attempts to start the pod.
5. That pod creation fails; in this case due to bad mount configuration
6. Kubernetes schedules that pod for cleanup
7. Kubernetes calls Calico via the CNI to clean the pod up
8. Calico removes the adapter & releases the IP
9. Everything is fine.

## Current Behavior

Assuming the same steps above, the problem is as so:

5. That pod creation fails
6. *Something* (I would assume docker) cleans up the veth pair for that container
7. Kubernetes calls Calico via the CNI to clean the container up
8. Calico determines somehow the adapter to delete *incorrectly*. In the logs below, you'll see the adapter that is deleted is not the one recorded against that pod object. 
9. Calico releases that IP. 
10. Calico *thinks* everything is fine. It is not.

Later, when that pod is eventually deleted Calico attempts to clean up that device, discovers it is not there and carries on ignoring that.

## Possible Solution

It would depend on the nature of the problem; I am *guessing* that the device id is not checked when the DEL command is issued; rather a proxy where the relevant device can be looked up — namespace path, for example. This might be where it goes awry, as the namespace path (if i understand correctly) is `/proc/$PID/ns/net`, and that `$PID` might get reallocated.

(These compute nodes are hella busy).

## Steps to Reproduce (for bugs)

Absolutely no idea. It happens rarely on production. The reconstruction above is from logs.

## Context

We have to put liveness probes on things to reboot them and, apparently, re-trigger the creation of these devices (or so). TBD — Still working on a hotfix.

## Your Environment
* Calico version

```
calico/node:v3.8.1
```

* Orchestrator version (e.g. kubernetes, mesos, rkt):

kubernetes v1.15.x

* Operating System and version:

Ubuntu 18.04

* Link to your project (optional):

(It's closed source)

## (Heavily Redacted) Logs

```

Aug 30 22:08:27 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:08:27.922 [INFO][11914] k8s.go 361: Populated endpoint ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0" Namespace="__NAMESPACE__" Pod="__POD_NAME__" WorkloadEndpoint="__WORKLOAD_ENDPOINT__" endpoint=&v3.WorkloadEndpoint{TypeMeta:v1.TypeMeta{Kind:"WorkloadEndpoint", APIVersion:"projectcalico.org/v3"}, ObjectMeta:v1.ObjectMeta{Name:"__WORKLOAD_ENDPOINT__", GenerateName:"__PARTIAL_POD_NAME__", Namespace:"__NAMESPACE__", SelfLink:"", UID:"8b9ffc8a-19f0-4f23-9fee-edf4cb8bc8e5", ResourceVersion:"475949471", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63734422103, loc:(*time.Location)(0x2346b20)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"projectcalico.org/orchestrator":"k8s", "projectcalico.org/serviceaccount":"__SERVICE_ACOUNT__", "app":"__APP_NAME__", "pod-template-hash":"6867d79cdb", "projectcalico.org/namespace":"__NAMESPACE__"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v3.WorkloadEndpointSpec{Orchestrator:"k8s", Workload:"", Node:"__NODE_NAME__", ContainerID:"", Pod:"__POD_NAME__", Endpoint:"eth0", IPNetworks:[]string{"10.217.51.215/32"}, IPNATs:[]v3.IPNAT(nil), IPv4Gateway:"", IPv6Gateway:"", Profiles:[]string{"kns.__NAMESPACE__", "ksa.__NAMESPACE__.__SERVICE_ACOUNT__"}, InterfaceName:"cali23af261a86a", MAC:"", Ports:[]v3.EndpointPort(nil)}}
Aug 30 22:08:27 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:08:27.922 [INFO][11914] k8s.go 362: Calico CNI using IPs: [10.217.51.215/32] ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0" Namespace="__NAMESPACE__" Pod="__POD_NAME__" WorkloadEndpoint="__WORKLOAD_ENDPOINT__"
Aug 30 22:08:27 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:08:27.922 [INFO][11914] network_linux.go 76: Setting the host side veth name to cali23af261a86a ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0" Namespace="__NAMESPACE__" Pod="__POD_NAME__" WorkloadEndpoint="__WORKLOAD_ENDPOINT__"
Aug 30 22:08:27 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:08:27.939 [INFO][11914] network_linux.go 396: Disabling IPv4 forwarding ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0" Namespace="__NAMESPACE__" Pod="__POD_NAME__" WorkloadEndpoint="__WORKLOAD_ENDPOINT__"
Aug 30 22:08:28 __NODE_NAME__ systemd-networkd[27141]: cali23af261a86a: Gained carrier
Aug 30 22:08:28 __NODE_NAME__ networkd-dispatcher[27494]: WARNING:Unknown index 1913 seen, reloading interface list
Aug 30 22:08:28 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:08:28.060 [INFO][11914] k8s.go 388: Added Mac, interface name, and active container ID to endpoint ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0" Namespace="__NAMESPACE__" Pod="__POD_NAME__" WorkloadEndpoint="__WORKLOAD_ENDPOINT__" endpoint=&v3.WorkloadEndpoint{TypeMeta:v1.TypeMeta{Kind:"WorkloadEndpoint", APIVersion:"projectcalico.org/v3"}, ObjectMeta:v1.ObjectMeta{Name:"__WORKLOAD_ENDPOINT__", GenerateName:"__PARTIAL_POD_NAME__", Namespace:"__NAMESPACE__", SelfLink:"", UID:"8b9ffc8a-19f0-4f23-9fee-edf4cb8bc8e5", ResourceVersion:"475949471", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63734422103, loc:(*time.Location)(0x2346b20)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"projectcalico.org/namespace":"__NAMESPACE__", "projectcalico.org/orchestrator":"k8s", "projectcalico.org/serviceaccount":"__SERVICE_ACOUNT__", "app":"__APP_NAME__", "pod-template-hash":"6867d79cdb"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v3.WorkloadEndpointSpec{Orchestrator:"k8s", Workload:"", Node:"__NODE_NAME__", ContainerID:"2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0", Pod:"__POD_NAME__", Endpoint:"eth0", IPNetworks:[]string{"10.217.51.215/32"}, IPNATs:[]v3.IPNAT(nil), IPv4Gateway:"", IPv6Gateway:"", Profiles:[]string{"kns.__NAMESPACE__", "ksa.__NAMESPACE__.__SERVICE_ACOUNT__"}, InterfaceName:"cali23af261a86a", MAC:"ae:09:0e:99:3e:54", Ports:[]v3.EndpointPort(nil)}}
...
Aug 30 22:10:01 __NODE_NAME__ systemd-networkd[27141]: cali23af261a86a: Lost carrier
...
Aug 30 22:10:03 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:10:03.659 [INFO][31825] plugin.go 442: Extracted identifiers ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0" Node="__NODE_NAME__" Orchestrator="k8s" WorkloadEndpoint="__WORKLOAD_ENDPOINT__"
Aug 30 22:10:03 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:10:03.677 [WARNING][31825] workloadendpoint.go 75: Operation Delete is not supported on WorkloadEndpoint type
Aug 30 22:10:03 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:10:03.677 [INFO][31825] k8s.go 470: Endpoint deletion will be handled by Kubernetes deletion of the Pod. ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0" endpoint=&v3.WorkloadEndpoint{TypeMeta:v1.TypeMeta{Kind:"WorkloadEndpoint", APIVersion:"projectcalico.org/v3"}, ObjectMeta:v1.ObjectMeta{Name:"__WORKLOAD_ENDPOINT__", GenerateName:"__PARTIAL_POD_NAME__", Namespace:"__NAMESPACE__", SelfLink:"", UID:"8b9ffc8a-19f0-4f23-9fee-edf4cb8bc8e5", ResourceVersion:"475951609", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63734422103, loc:(*time.Location)(0x2346b20)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"app":"__APP_NAME__", "pod-template-hash":"6867d79cdb", "projectcalico.org/namespace":"__NAMESPACE__", "projectcalico.org/orchestrator":"k8s", "projectcalico.org/serviceaccount":"__SERVICE_ACOUNT__"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v3.WorkloadEndpointSpec{Orchestrator:"k8s", Workload:"", Node:"__NODE_NAME__", ContainerID:"", Pod:"__POD_NAME__", Endpoint:"eth0", IPNetworks:[]string{"10.217.51.215/32"}, IPNATs:[]v3.IPNAT(nil), IPv4Gateway:"", IPv6Gateway:"", Profiles:[]string{"kns.__NAMESPACE__", "ksa.__NAMESPACE__.__SERVICE_ACOUNT__"}, InterfaceName:"cali23af261a86a", MAC:"", Ports:[]v3.EndpointPort(nil)}}
Aug 30 22:10:03 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:10:03.677 [INFO][31825] k8s.go 477: Cleaning up netns ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0"
Aug 30 22:10:03 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:10:03.682 [INFO][31825] network_linux.go 450: Calico CNI deleting device in netns /proc/1316/ns/net ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0"
Aug 30 22:10:03 __NODE_NAME__ systemd-networkd[27141]: calicd9e81b465b: Lost carrier
Aug 30 22:10:03 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:10:03.804 [INFO][31825] network_linux.go 467: Calico CNI deleted device in netns /proc/1316/ns/net ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0"
Aug 30 22:10:03 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:10:03.804 [INFO][31825] k8s.go 484: Releasing IP address(es) ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0"
Aug 30 22:10:03 __NODE_NAME__ kubelet[27693]: 2020-08-30 22:10:03.804 [INFO][31825] utils.go 168: Calico CNI releasing IP address ContainerID="2a0894d5fbf6d033c94cab8bb328e45762ce63bd69678fd16953e58ac6482cc0"
```

**Comment by andrewhowdencom** [comment]
(calicd9e81b465b is the one that shouldn't have been deleted, cali23af261a86a is the one that was / should have been)

**Comment by caseydavenport** [comment]
Yeah, the Calico CNI plugin will use the netns+ifce name in order to find the correct interface to clean up on teardown (https://github.com/projectcalico/cni-plugin/blob/master/pkg/dataplane/linux/dataplane_linux.go#L462). 

But, it's trusting that the container runtime is passing it the correct netns and interface name. I don't think we can take the arguments given to us by the runtime and say "actually, I think you meant to delete this interface!"

Basically, I don't see how this would be a result of the pid being re-used, unless it was a bug in the container runtime itself, since the container runtime shouldn't send us DEL requests for the wrong netns. Does that make sense? 

What container runtime are you guys using?

**Comment by song-jiang** [comment]
@andrewhowdencom Would you provide more info on this?

**Comment by andrewhowdencom** [comment]
Broadly, it happens fairly rarely so we created a liveness probe that probed TCP and forced a reboot (and correspondingly reconstruction of the adapter) and the problem is papered over.

**Comment by andrewhowdencom** [comment]
(to answer the earlier question, docker)

**Comment by caseydavenport** [comment]
> (to answer the earlier question, docker)

Hm, we test pretty extensively with Docker, but I suppose that an issue like this might just appear as a flake and get missed. I'm still feeling confident that this is a runtime / configuration issue rather than a Calico issue. The CNI plugin needds to assume the netns it is provided is correct. If it's given an incorrect netns... well, there's not much it can do!
